File: xpod/extensions/NSTask+XPPod.m
//
//  NSTask+XPPod.m
//  xpod
//
//  Created by Maxime Bokobza on 02/06/14.
//  Copyright (c) 2014 dispatchasync. All rights reserved.
//

#import "NSTask+XPPod.h"

@implementation NSTask (XPPod)

+ (NSString *)launchTaskWithLaunchPath:(NSString *)launchPath arguments:(NSArray *)arguments {
    return [self launchTaskFromDirectory:nil withLaunchPath:launchPath arguments:arguments];
}

+ (NSString *)launchTaskFromDirectory:(NSString *)currentDirectory withLaunchPath:(NSString *)launchPath arguments:(NSArray *)arguments {
    NSTask *task = [[NSTask alloc] init];
    if (currentDirectory) {
        [self createDirectory:currentDirectory];
        task.currentDirectoryPath = currentDirectory;
    }
    task.launchPath = launchPath;
    task.arguments = arguments;
    
    NSPipe *output = [[NSPipe alloc] init];
    [task setStandardOutput:output];
    
    [task launch];
    [task waitUntilExit];
    
    NSData *outputData = [[output fileHandleForReading] readDataToEndOfFile];
    return [[NSString alloc] initWithData:outputData encoding:NSUTF8StringEncoding];
}

+ (NSArray *)filenamesInDirectory:(NSString *)directory ofType:(NSString *)extension {
    NSString *output = [self launchTaskWithLaunchPath:@"/bin/ls"
                                            arguments:@[directory]];
    
    NSMutableArray *filenames = [NSMutableArray array];
    for (NSString *filename in [output componentsSeparatedByString:@"\n"]) {
        if ([filename hasSuffix:[NSString stringWithFormat:@".%@", extension]])
            [filenames addObject:[directory stringByAppendingPathComponent:filename]];
    }
    
    return [NSArray arrayWithArray:filenames];
}

+ (void)createDirectory:(NSString *)directory {
    if ([[NSFileManager defaultManager] fileExistsAtPath:directory])
        return;
    
    [self launchTaskWithLaunchPath:@"/bin/mkdir"
                         arguments:@[@"-p", directory]];
}

@end

File: xpod/extensions/NSTask+XPPod.swift
//
//  NSTask+XPPod.swift
//  xpod
//
//  Created by Maxime Bokobza on 12/06/14.
//  Copyright (c) 2014 dispatchasync. All rights reserved.
//

import Foundation


extension NSTask {
    class func launchTask(directory: String?, launchPath: String, arguments: String[]) -> String {
        var task = NSTask()
        if directory != nil {
            self.createDirectory(directory!)
            task.currentDirectoryPath = directory;
        }
        
        task.launchPath = launchPath
        task.arguments = arguments
        
        var output = NSPipe();
        task.standardOutput = output;
        
        task.launch()
        task.waitUntilExit()
        
        var outputData = output.fileHandleForReading.readDataToEndOfFile()
        return NSString(data: outputData, encoding: NSUTF8StringEncoding)
    }
    
    class func launchTask(launchPath: String, arguments: String[]) -> String {
        return self.launchTask(nil, launchPath: launchPath, arguments: arguments)
    }
    
    class func createDirectory(directory: String) {
        if NSFileManager.defaultManager().fileExistsAtPath(directory) {
            return;
        }
        
        NSFileManager.defaultManager().createDirectoryAtPath(directory, withIntermediateDirectories: true, attributes: nil, error: nil)
    }
    
    class func listFiles(directory: String, fileType: String) -> String[] {
        var files = NSFileManager.defaultManager().contentsOfDirectoryAtPath(directory, error: nil) as String[]
        
        var filenames = String[]();
        for filename in files {
            if filename.hasSuffix(".\(fileType)") {
                filenames.append(directory.stringByAppendingPathComponent(filename))
            }
        }
        
        return filenames
    }
    
    class func copy(files: String[], destination: String) {
        self.createDirectory(destination)
        
        for file in files {
            let path = destination + "/" + file.lastPathComponent
            NSFileManager.defaultManager().copyItemAtPath(file, toPath: path, error: nil)
        }
    }
    
    class func symlink(aliasPath: String, destinationPath: String) {
        println(aliasPath)
        println(destinationPath)
        NSFileManager.defaultManager().createSymbolicLinkAtPath(aliasPath, withDestinationPath: destinationPath, error: nil)
    }
}
